File: backend/internal/services/catalog/scene_week.go
package catalog

import (
	"fmt"
	"strconv"
	"strings"
	"time"

	apperrors "psychic-homily-backend/internal/errors"
	catalogm "psychic-homily-backend/internal/models/catalog"
	"psychic-homily-backend/internal/services/contracts"
	"psychic-homily-backend/internal/services/shared"
	"psychic-homily-backend/internal/utils"
)

// sceneWeekShowCap bounds one week's payload. Chicago, the densest scene,
// runs ~50 shows in a busy week, so this leaves comfortable headroom while
// still refusing to serve an unbounded list.
const sceneWeekShowCap = 300

// ISOWeekKey formats a time as its ISO-8601 week key ("2026-W31"). The ISO year
// is NOT always the calendar year — 2027-01-01 belongs to ISO week 2026-W53 —
// so the year must come from ISOWeek(), never from Year().
func ISOWeekKey(t time.Time) string {
	y, w := t.ISOWeek()
	return fmt.Sprintf("%04d-W%02d", y, w)
}

// ParseISOWeekKey parses "2026-W31" into its ISO year and week number.
//
// Validation is strict because these values arrive from URLs: the week must be
// 1..53 AND must actually exist in that year. Most years have 52 weeks, so
// "2026-W53" may be a real week or may not; the round-trip check below is what
// distinguishes them, and it is the reason this cannot be a bare regex.
func ParseISOWeekKey(key string, loc *time.Location) (year, week int, err error) {
	parts := strings.Split(strings.ToUpper(strings.TrimSpace(key)), "-W")
	if len(parts) != 2 {
		return 0, 0, fmt.Errorf("malformed ISO week key %q (want YYYY-Www)", key)
	}
	year, err = strconv.Atoi(parts[0])
	if err != nil || year < 1970 || year > 9999 {
		return 0, 0, fmt.Errorf("malformed ISO week year in %q", key)
	}
	week, err = strconv.Atoi(parts[1])
	if err != nil || week < 1 || week > 53 {
		return 0, 0, fmt.Errorf("ISO week out of range in %q", key)
	}
	// Round-trip: a 53rd week in a 52-week year lands in the NEXT ISO year, so
	// the computed Monday disagrees with the requested key.
	start := ISOWeekStart(year, week, loc)
	gotY, gotW := start.ISOWeek()
	if gotY != year || gotW != week {
		return 0, 0, fmt.Errorf("ISO week %q does not exist", key)
	}
	return year, week, nil
}

// ISOWeekStart returns midnight on the Monday opening the given ISO week, in
// loc.
//
// Anchored on January 4th, which ISO-8601 guarantees falls in week 1 of its
// year. Anchoring on January 1st would be wrong: it can belong to the last week
// of the PREVIOUS ISO year.
func ISOWeekStart(year, week int, loc *time.Location) time.Time {
	if loc == nil {
		loc = time.UTC
	}
	jan4 := time.Date(year, time.January, 4, 0, 0, 0, 0, loc)
	// Go weekdays run Sunday=0..Saturday=6; ISO runs Monday=1..Sunday=7.
	weekday := int(jan4.Weekday())
	if weekday == 0 {
		weekday = 7
	}
	week1Monday := jan4.AddDate(0, 0, -(weekday - 1))
	return week1Monday.AddDate(0, 0, (week-1)*7)
}

// sceneCalendarWeekWindow returns the half-open [start, end) instants bounding
// the Monday-Sunday week that CONTAINS now, in loc.
//
// The same two instants GetSceneWeek derives for its current week, reached
// through the same ISOWeekStart: the calendar maths lives in exactly one place,
// so a list count and the page it links to cannot drift onto different Mondays.
// end is walked forward by seven CALENDAR days rather than 168 hours, so a week
// containing a DST transition still closes at local midnight.
func sceneCalendarWeekWindow(now time.Time, loc *time.Location) (start, end time.Time) {
	if loc == nil {
		loc = time.UTC
	}
	year, week := now.In(loc).ISOWeek()
	start = ISOWeekStart(year, week, loc)
	return start, start.AddDate(0, 0, 7)
}

// sceneTimezonesByKey resolves every scene's week-boundary timezone in ONE
// query, keyed the way ListScenes groups venues.
//
// The modal rule is sceneLocation's, restated over all scenes at once rather
// than per scene: most common explicit timezone among the scene's VERIFIED
// venues, ties broken alphabetically (the ROW_NUMBER ordering reproduces that
// query's ORDER BY, over the same venue rows).
//
// SCOPING IS NOT IDENTICAL TO sceneScope.venuePredicate, and the difference is
// worth knowing before trusting an equality here:
//   - METRO scene: exact. Its key IS the CBSA, so COALESCE(v.metro, …) = key
//     selects the same rows as `v.metro = ?`. CBSA codes are numeric and
//     fallback keys always contain '|', so the two key spaces cannot collide.
//   - FALLBACK (no-CBSA) scene: a SUBSET. This key requires v.metro IS NULL;
//     venuePredicate matches the (city, state) whether or not the venue has a
//     metro. They agree only while every venue in that place agrees about its
//     metro — which is what the geocoder writes, but venues.metro DRIFTS (see
//     metro_backfill.go: the background location writers move city/state
//     without touching metro). A place holding both NULL-metro and CBSA venues
//     already splits into two ListScenes groups sharing one slug; that is
//     pre-existing, and it is the case where this count can undercount the
//     week page.
//
// Scenes missing from the returned map have no verified venue carrying a
// timezone; the caller falls back to the state map exactly as sceneLocation
// does.
func (s *SceneService) sceneTimezonesByKey() (map[string]string, error) {
	if s.db == nil {
		return nil, fmt.Errorf("database not initialized")
	}
	type tzRow struct {
		SceneKey string `gorm:"column:scene_key"`
		Timezone string `gorm:"column:timezone"`
	}
	var rows []tzRow
	// The window function ranks each scene's timezones by venue count; the
	// outer WHERE keeps the winner. A NULL key (a venue with no metro and no
	// usable city/state) can match no scene, so it is dropped here rather than
	// scanning into an empty-string key that looks like a real one.
	err := s.db.Raw(`
		SELECT scene_key, timezone FROM (
			SELECT ` + sceneGroupKeySQL + ` AS scene_key,
			       v.timezone AS timezone,
			       ROW_NUMBER() OVER (
			           PARTITION BY ` + sceneGroupKeySQL + `
			           ORDER BY COUNT(*) DESC, v.timezone ASC
			       ) AS rn
			FROM venues v
			WHERE v.verified = true
			  AND v.timezone IS NOT NULL
			  AND v.timezone <> ''
			GROUP BY ` + sceneGroupKeySQL + `, v.timezone
		) ranked
		WHERE rn = 1 AND scene_key IS NOT NULL
	`).Scan(&rows).Error
	if err != nil {
		return nil, fmt.Errorf("failed to resolve scene timezones: %w", err)
	}
	out := make(map[string]string, len(rows))
	for _, r := range rows {
		out[r.SceneKey] = r.Timezone
	}
	return out, nil
}

// sceneCalendarWeekTarget is one scene to count a calendar week for: the key
// ListScenes grouped it under, plus the display state that backs the timezone
// fallback.
type sceneCalendarWeekTarget struct {
	key   string
	state string
}

// sceneCalendarWeekCounts returns, per scene key, how many shows that scene's
// /scenes/{slug}/week page reports for the week containing now.
//
// Two queries for the whole list, not two per scene: one resolves every
// scene's timezone, one counts every scene's own week. The count query joins a
// VALUES list of per-scene windows, which is what lets a single statement apply
// a DIFFERENT Monday to Chicago than to Honolulu.
//
// The predicate is GetSceneShowsInRange's, deliberately down to its omissions:
// approved shows at ANY venue in scope, verified or not, cancelled ones
// included. This number's only job is to equal the destination page's total, so
// it has to be counted the way that page counts, not the way that page ideally
// would. Capped at sceneWeekShowCap for the same reason — the page's total is
// the length of a capped list, so an uncapped count would overstate it for a
// scene busy enough to hit the ceiling.
//
// CONSEQUENCE, because it will look like a bug from the outside: this counts a
// DIFFERENT venue population from every other number on SceneListResponse. The
// grouped ListScenes query applies sceneVenueEligibilitySQL (verified venues
// with a usable city/state), so venue_count, total_show_count,
// upcoming_show_count and shows_this_week are verified-only. A scene with
// unverified rooms can therefore read "3 shows · 17 shows this week". Adding a
// verified filter here would make the card self-consistent and make the LINK
// wrong, which is the trade this field exists to refuse. Fix it, if it is worth
// fixing, by narrowing the week PAGE — then this follows for free.
//
// Grouping caveat: see sceneTimezonesByKey. The metro branch is exact; a
// fallback scene whose place holds both NULL-metro and CBSA venues can
// undercount.
func (s *SceneService) sceneCalendarWeekCounts(now time.Time, targets []sceneCalendarWeekTarget) (map[string]int, error) {
	if s.db == nil {
		return nil, fmt.Errorf("database not initialized")
	}
	if len(targets) == 0 {
		return map[string]int{}, nil
	}

	timezones, err := s.sceneTimezonesByKey()
	if err != nil {
		return nil, err
	}

	values := make([]string, 0, len(targets))
	args := make([]any, 0, len(targets)*3+1)
	for _, t := range targets {
		var loc *time.Location
		if tz, ok := timezones[t.key]; ok && tz != "" {
			loc = utils.EventLocation(&tz, t.state)
		} else {
			loc = utils.EventLocation(nil, t.state)
		}
		start, end := sceneCalendarWeekWindow(now, loc)
		// Casts on every row, not just the first: without them Postgres types
		// the VALUES columns as `unknown` and the join comparison fails.
		values = append(values, "(?::text, ?::timestamptz, ?::timestamptz)")
		args = append(args, t.key, start.UTC(), end.UTC())
	}
	args = append(args, catalogm.ShowStatusApproved)

	type countRow struct {
		SceneKey string `gorm:"column:scene_key"`
		Count    int    `gorm:"column:count"`
	}
	var rows []countRow
	if err := s.db.Raw(`
		SELECT w.scene_key AS scene_key,
		       LEAST(COUNT(DISTINCT s.id), `+strconv.Itoa(sceneWeekShowCap)+`) AS count
		FROM (VALUES `+strings.Join(values, ", ")+`) AS w(scene_key, wk_start, wk_end)
		JOIN venues v ON `+sceneGroupKeySQL+` = w.scene_key
		JOIN show_venues sv ON sv.venue_id = v.id
		JOIN shows s ON s.id = sv.show_id
		WHERE s.status = ?
		  AND s.event_date >= w.wk_start
		  AND s.event_date < w.wk_end
		GROUP BY w.scene_key
	`, args...).Scan(&rows).Error; err != nil {
		return nil, fmt.Errorf("failed to count scene calendar weeks: %w", err)
	}

	out := make(map[string]int, len(rows))
	for _, r := range rows {
		out[r.SceneKey] = r.Count
	}
	return out, nil
}

// weekHasEnded reports whether a scene's week is entirely behind it.
//
// end is the week's EXCLUSIVE end — the same instant that closes the show
// query's half-open window — so this asks "has the scene's own clock reached
// the following Monday". Both arguments are absolute instants, which is what
// keeps it right across a DST transition: end was walked forward by calendar
// days rather than by 168 hours, so a week containing a transition still ends
// at local midnight rather than an hour either side of it.
//
// False for the current week and for every future week alike. That is the whole
// distinction the caller needs and the reason this is not simply
// !IsCurrentWeek: only a week that can no longer gain shows is safe to freeze.
func weekHasEnded(now, end time.Time) bool {
	return !now.Before(end)
}

// sceneLocation resolves the timezone a scene's week boundaries are computed
// in: the most common explicit timezone among its verified venues, falling back
// to the state map, both resolved through shared.EventZone.
//
// Modal rather than first-row so one mis-geocoded venue cannot shift the whole
// city's week. Metro scopes stay within a single zone in practice (a CBSA is
// geographically compact), so a single location per scene is well defined.
//
// The second return is the publishable zone NAME, nil when neither the modal
// venue zone nor the state map answered and the location is the Arizona
// default. Both are returned together so a caller cannot compute the boundaries
// on one input and publish a zone derived from another.
func (s *SceneService) sceneLocation(scope sceneScope, state string) (*time.Location, *string) {
	if s.db == nil {
		return shared.EventZone(nil, state)
	}
	vp, vargs := scope.venuePredicate("v")
	var tz string
	err := s.db.Raw(`
		SELECT v.timezone
		FROM venues v
		WHERE `+vp+`
		  AND v.verified = true
		  AND v.timezone IS NOT NULL
		  AND v.timezone <> ''
		GROUP BY v.timezone
		ORDER BY COUNT(*) DESC, v.timezone ASC
		LIMIT 1
	`, vargs...).Scan(&tz).Error
	if err != nil || tz == "" {
		return shared.EventZone(nil, state)
	}
	return shared.EventZone(&tz, state)
}

// GetSceneWeek returns one ISO week of a scene's shows, grouped by day in the
// scene's own timezone.
//
// weekKey is either an ISO week key ("2026-W31") or "" for the scene's current
// week. Resolving "current" here rather than on the client is deliberate: the
// answer depends on the SCENE's timezone, not the viewer's, so a reader in
// Berlin and a reader in Chicago must get the same Chicago week.
//
// A valid scene with no shows in the window returns a populated response with
// zero shows, not an error — a real city having a quiet week is a fact, and
// 404ing it would make an already-shared permalink break retroactively. Only an
// unknown or below-threshold scene is a 404.
func (s *SceneService) GetSceneWeek(city, state, weekKey string) (*contracts.SceneWeekResponse, error) {
	if s.db == nil {
		return nil, fmt.Errorf("database not initialized")
	}

	scope, err := s.scopeFor(city, state)
	if err != nil {
		return nil, err
	}
	if n, err := s.verifiedVenueCount(scope); err != nil {
		return nil, fmt.Errorf("failed to count venues: %w", err)
	} else if n < sceneMinVenues {
		return nil, apperrors.ErrSceneNotFound(fmt.Sprintf("scene not found: %s, %s", city, state))
	}

	loc, zone := s.sceneLocation(scope, state)
	nowLocal := time.Now().In(loc)
	currentKey := ISOWeekKey(nowLocal)

	if strings.TrimSpace(weekKey) == "" {
		weekKey = currentKey
	}
	year, week, err := ParseISOWeekKey(weekKey, loc)
	if err != nil {
		return nil, apperrors.ErrSceneNotFound(err.Error())
	}

	start := ISOWeekStart(year, week, loc)
	end := start.AddDate(0, 0, 7) // half-open [start, end)

	shows, err := s.GetSceneShowsInRange(city, state, start.UTC(), end.UTC(), loc, sceneWeekShowCap)
	if err != nil {
		return nil, err
	}

	// Pre-seed all seven days so quiet nights render as empty rather than
	// vanishing, then bucket by the scene-local date the query already
	// formatted.
	// Seeded with empty (non-nil) slices so a quiet day marshals as `[]` rather
	// than `null`. A nil slice would force every client to handle two shapes
	// for "no shows", and a TS type declaring `shows: Show[]` would silently
	// receive null.
	byDate := make(map[string][]contracts.SceneShowSummary, 7)
	days := make([]contracts.SceneWeekDay, 0, 7)
	for i := 0; i < 7; i++ {
		key := start.AddDate(0, 0, i).Format("2006-01-02")
		byDate[key] = []contracts.SceneShowSummary{}
		days = append(days, contracts.SceneWeekDay{Date: key})
	}
	for _, sh := range shows {
		if _, ok := byDate[sh.EventDate]; ok {
			byDate[sh.EventDate] = append(byDate[sh.EventDate], sh)
		}
	}
	for i := range days {
		days[i].Shows = byDate[days[i].Date]
	}

	venues, err := s.trackedVenueDetails(scope)
	if err != nil {
		return nil, err
	}
	if venues == nil {
		venues = []contracts.SceneTrackedVenue{}
	}

	return &contracts.SceneWeekResponse{
		// Canonical slug, not whatever the caller asked for: a metro MEMBER
		// slug (mesa-az) resolves to its principal city (phoenix-az), and the
		// client builds prev/next week URLs from this.
		Slug:          buildSceneSlug(city, state),
		SceneName:     fmt.Sprintf("%s, %s", city, state),
		City:          city,
		State:         state,
		ISOWeek:       ISOWeekKey(start),
		StartDate:     start.Format("2006-01-02"),
		EndDate:       start.AddDate(0, 0, 6).Format("2006-01-02"),
		Timezone:      zone,
		ShowCount:     len(shows),
		PrevWeek:      ISOWeekKey(start.AddDate(0, 0, -7)),
		NextWeek:      ISOWeekKey(start.AddDate(0, 0, 7)),
		IsCurrentWeek: ISOWeekKey(start) == currentKey,
		IsPastWeek:    weekHasEnded(nowLocal, end),
		Days:          days,
		TrackedVenues: venues,
	}, nil
}

File: frontend/features/scenes/sceneWeek.ts
import type { components } from '@/types/api'

/**
 * Scene-week types are DERIVED from the generated OpenAPI schema, not
 * hand-written.
 *
 * The rest of this feature predates `types/api.d.ts` and hand-writes its
 * interfaces, which is the exact drift hazard that has bitten this codebase
 * before: a hand-written type can claim a field the API never sends, and the
 * mismatch is invisible to CI while being fatal in production. New surfaces
 * derive from the schema so the `api:types:check` drift gate keeps them honest.
 */
export type SceneWeekResponse = components['schemas']['SceneWeekResponse']
export type SceneWeekDay = components['schemas']['SceneWeekDay']
export type SceneWeekShow = components['schemas']['SceneShowSummary']

/** ISO week key, e.g. `2026-W31`. Case-insensitive, zero-padded week. */
const ISO_WEEK_KEY = /^\d{4}-W\d{2}$/i

/**
 * Earliest week key worth serving. The site has no shows before this, and an
 * unbounded key space is an unbounded set of distinct URLs — each one a cache
 * miss that renders a full share card. Bounding it keeps that finite.
 */
const FIRST_TRACKED_YEAR = 2015

/**
 * Whether a URL segment even looks like an ISO week key.
 *
 * A cheap shape check only — it deliberately does NOT decide whether the week
 * exists. `2025-W53` is well-formed but not a real week (2025 has 52), and only
 * the backend, which owns the calendar maths and the scene's timezone, can say
 * so. This exists to avoid a pointless round-trip for obvious junk.
 *
 * Matches the RAW segment, deliberately. Trimming here would accept forms the
 * proxy's copy of this rule rejects, so a page and its share card would
 * disagree about which URLs are legal — and every accepted variant of the same
 * week (leading space, tab, BOM) is another distinct URL that renders its own
 * card. One spelling per week.
 */
export function looksLikeISOWeek(segment: string): boolean {
  const match = ISO_WEEK_KEY.exec(segment)
  if (!match) return false
  const year = Number(segment.slice(0, 4))
  return year >= FIRST_TRACKED_YEAR && year <= new Date().getUTCFullYear() + 1
}

/**
 * Resolve a requested week segment into what should actually be fetched.
 *
 * Three outcomes, and the difference between the last two is the whole point:
 * - `undefined` — no segment supplied; the caller wants the CURRENT week.
 * - a normalized key — a well-formed week to fetch.
 * - `null` — the segment is junk and NOTHING should be fetched.
 *
 * The archived route's segment is dynamic, so it also catches any unmatched
 * child path under a scene. Collapsing junk to `undefined` would quietly serve
 * the current week for a URL whose page 404s — handing out a confident-looking
 * answer to a question nobody asked. The key is returned exactly as validated,
 * so the string that was checked is the string that gets used.
 */
export function resolveRequestedWeek(segment: string | undefined): string | undefined | null {
  if (segment === undefined) return undefined
  return looksLikeISOWeek(segment) ? segment : null
}

/**
 * A show count as a sentence, optionally suffixed "this week".
 *
 * WHICH week is the CALLER's to know — this function only spells the phrase.
 * Every caller now passes a Monday-to-Sunday total in the scene's own timezone:
 * the share card reads it off the week payload, the scene cards and the
 * `/shows` by-city index read `shows_calendar_week` from `GET /scenes`. The
 * rolling `shows_this_week` is a different seven days and must not be spelled
 * with these words next to a link to a week page.
 *
 * The suffix is dropped for an archived share card, which carries its date
 * range directly above this line: that reads correctly for a week shared months
 * later rather than claiming to be current. A quiet week says so in words —
 * `0 shows` is a bad thing to post.
 */
export function formatShowCountLine(total: number, isCurrentWeek: boolean): string {
  const period = isCurrentWeek ? ' this week' : ''
  if (total === 0) return `No shows${period}`
  return `${total} ${total === 1 ? 'show' : 'shows'}${period}`
}

/**
 * Parse a `YYYY-MM-DD` date as LOCAL midnight.
 *
 * `new Date('2026-07-27')` parses as UTC midnight, which renders as Jul 26 in
 * any negative-offset timezone — so a US reader would see every day of the week
 * shifted back by one. The backend already resolved these dates in the scene's
 * own timezone; they are calendar dates, not instants, and must be built
 * component-wise to stay that way.
 *
 * Exported for the day surface, which parses the same scene-local ISO dates and
 * must not grow a second copy of this rule — a page and its sibling disagreeing
 * about what `2026-07-31` means is exactly the bug this guards.
 */
export function parseCalendarDate(iso: string): Date {
  const [y, m, d] = iso.split('-').map(Number)
  return new Date(y, (m ?? 1) - 1, d ?? 1)
}

/**
 * The zone the cross-city week index names its week in.
 *
 * It names a heading, and that is a different question from resolving
 * "upcoming", which is decided per show against its own venue's zone and needs
 * no shared viewer zone at all (PSY-1678). There is no per-row zone to move
 * this to: the heading sits above every row at once.
 *
 * A single zone is a compromise here, not a correct answer. The per-row counts
 * are each scene's own calendar week, exact against the page the row links to
 * (`shows_calendar_week`, PSY-1623); this heading is not, because the approved
 * mock puts one range above every row. The scene-week pages resolve "current
 * week" in each scene's OWN venue timezone, so for a few hours after Monday
 * midnight a scene east of this zone has already turned over while this label
 * still names the previous week — its count and its destination agree with each
 * other, and only the shared heading above them is stale.
 *
 * It is not the block's only approximation: the counts also reach the page
 * through a cache while this label is derived from a live clock, so they can lag
 * it at the rollover (see `ThisWeekByCity`).
 *
 * Fixing either properly means per-row ranges, which is a design change rather
 * than a data one: `GET /scenes` would also have to publish each scene's bounds,
 * and the heading would be read off the payload instead of the clock.
 */
export const SCENE_WEEK_INDEX_TIMEZONE = 'America/Los_Angeles'

/** `2026-07-27` from a `Date` read in its own local fields. */
function toCalendarDate(date: Date): string {
  const pad = (n: number) => String(n).padStart(2, '0')
  return `${date.getFullYear()}-${pad(date.getMonth() + 1)}-${pad(date.getDate())}`
}

/**
 * The Monday and Sunday bounding the week that CONTAINS `now` in `timeZone`.
 *
 * The scene-week pages get their bounds from the backend, which resolves them
 * in each scene's own zone. Nothing carries those bounds for a list of scenes,
 * so a surface that names one week across many cities has to derive it — and
 * derive it in a stated zone, because "which week is it" changes at a different
 * instant in Phoenix than in London.
 *
 * `formatToParts` reads the year, month and day the zone is currently on
 * without any date maths, and without depending on how a locale happens to
 * order a short date (`lib/utils/timeUtils.ts` and `lib/utils/showTiming.ts`
 * take the same route to the same question). The rest is plain local-field
 * arithmetic on a local-midnight `Date`, which `Date` normalizes across month
 * and year ends for us.
 *
 * `now` is a parameter rather than a `new Date()` inside, so tests can pin a
 * week and so the caller is forced to own reading the clock. Under
 * `cacheComponents` that read is only legal at request time.
 */
export function currentWeekBounds(
  now: Date,
  timeZone: string,
): { start: string; end: string } {
  const parts = new Intl.DateTimeFormat('en-US', {
    timeZone,
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
  }).formatToParts(now)
  const field = (type: string) => Number(parts.find(p => p.type === type)?.value)

  const today = new Date(field('year'), field('month') - 1, field('day'))
  // `getDay()` is Sunday-first; the week here is Monday-first, matching the ISO
  // week the `/scenes/{slug}/week` pages serve.
  const daysSinceMonday = (today.getDay() + 6) % 7
  const start = new Date(
    today.getFullYear(),
    today.getMonth(),
    today.getDate() - daysSinceMonday,
  )
  const end = new Date(
    start.getFullYear(),
    start.getMonth(),
    start.getDate() + 6,
  )
  return { start: toCalendarDate(start), end: toCalendarDate(end) }
}

/** `JUL 27` — the shared stem of every uppercase date label here. */
function monthDay(iso: string): string {
  const date = parseCalendarDate(iso)
  const month = date.toLocaleDateString('en-US', { month: 'short' })
  return `${month} ${date.getDate()}`.toUpperCase()
}

/** `MON JUL 27` — the day-group heading. */
export function formatDayHeading(iso: string): string {
  const weekday = parseCalendarDate(iso).toLocaleDateString('en-US', { weekday: 'short' })
  return `${weekday.toUpperCase()} ${monthDay(iso)}`
}

/**
 * `JUL 27 – AUG 2` — the share card's week range.
 *
 * Drops the weekday and year the page header carries: on a card the range sits
 * above the city at display scale and only has to say *which week*, and at the
 * 300px a link renders at in a group chat every dropped word buys size on the
 * words that remain.
 */
export function formatWeekRangeCompact(startISO: string, endISO: string): string {
  return `${monthDay(startISO)} – ${monthDay(endISO)}`
}

/** `Mon Jul 27 – Sun Aug 2, 2026` — the header's week range. */
export function formatWeekRange(startISO: string, endISO: string): string {
  const start = parseCalendarDate(startISO)
  const end = parseCalendarDate(endISO)
  const fmt = (d: Date) =>
    d.toLocaleDateString('en-US', { weekday: 'short', month: 'short', day: 'numeric' })
  return `${fmt(start)} – ${fmt(end)}, ${end.getFullYear()}`
}

/**
 * The line a reader actually scans.
 *
 * Most shows carry an empty `title` — display names are composed from the bill
 * everywhere else in the app — so artist names are the primary source and the
 * title is the fallback, not the other way round.
 *
 * Shared with the day surface, which lists the same rows: the two pages must
 * name a show identically or a reader following one to the other sees it
 * rename itself.
 */
export function showDisplayTitle(show: SceneWeekShow): string {
  const names = show.artist_names ?? []
  if (names.length > 0) return names.join(', ')
  if (show.title) return show.title
  return 'Live music'
}

/** Canonical `/shows/...` target; falls back to the id when a slug is missing. */
export function showHref(show: SceneWeekShow): string {
  return show.slug ? `/shows/${show.slug}` : `/shows/${show.id}`
}

/**
 * Total shows actually listed.
 *
 * Prefer the server's `show_count`, but fall back to counting: `days` is typed
 * nullable by the generator even though the API always emits an array, and a
 * header that disagrees with the list below it is worse than a recount.
 */
export function countShows(week: SceneWeekResponse): number {
  if (typeof week.show_count === 'number') return week.show_count
  return (week.days ?? []).reduce((n, day) => n + (day.shows?.length ?? 0), 0)
}
